Update dependencies and code to php8.3

geneowak opened the pull request:
## What does the PR Do?
It updates all the composer dependencies and upgrades the code to php8.3 and fixes the deprecated code.

## Type of Change
- [ ] Bug fix
- [ ] New feature
- [ ] Documentation update
- [x] Refactoring

## Description
Many of the packages in the code have long been updated and provided with new features. PHP has also made a number of advances and the code was due for an update.
The following is what has been done
- Updated all composer dependencies to the latest versions
- Installed php rector and used it to update the code to php8.3
- Added laravel pint for lint
- Did some tests to make sure that the code still works as intended

## Testing
- Checkout the branch/repo `git clone git@github.com:geneowak/laracasts-downloader.git`
- Run composer install
- Add a .env file and try downloading episode/series

## Impact
Not sure if this is something the author might be interested in but thought that we could benefit from the new updates to php and the composer packages.

geneowak committed:
Update to latest

- update packages to latest
- require php8.3
- add php rector and configure it

geneowak committed:
Refactor code to latest php

- run phprector to update code to php8.3
- add pint and lint code
- refactor some deprecated code e.g dotenv stuff...

geneowak committed:
Update outdated and deprecated code in guzzle and the filesystem.

geneowak committed:
Use sink instead of save_to in guzzle.

geneowak committed:
Update docker php version to 8.3. and update the readme.

carlosflorencio commented:
@geneowak thanks for the contribution! Looks good to me :).

Will wait for @AmirRezaM75 comment as well.

AmirRezaM75 commented:
I will take a look at it on Saturday. thanks

AmirRezaM75 commented on App/Html/Parser.php:
```diff
@@ -134,7 +128,7 @@ public static function extractLarabitsSeries($html)
     public static function getCsrfToken($html)
     {
-        preg_match('/"csrfToken": \'([^\s]+)\'/', $html, $matches);
+        preg_match('/"csrfToken": \'([^\s]+)\'/', (string) $html, $matches);
```
please use argument type hint instead of casting.

```php
public static function getCsrfToken(string $html)
```

AmirRezaM75 commented on App/Html/Parser.php:
```diff
@@ -125,7 +119,7 @@ public static function getEpisodeDownloadLink($episodeHtml)
     {
-        $html = str_replace('\/', '/', html_entity_decode($html));
+        $html = str_replace('\/', '/', html_entity_decode((string) $html));
 
```
please use argument type hint instead of casting.

```php
public static function extractLarabitsSeries(string $html)
```

AmirRezaM75 commented on App/Vimeo/VimeoDownloader.php:
```diff
@@ -63,22 +61,19 @@ private function downloadSource($baseURL, $sourceData, $filepath)
     private function downloadSegments($segmentURLs, $filepath, $sizes)
     {
-        $type = strpos($filepath, 'm4v') !== false ? 'video' : 'audio';
+        $type = str_contains((string) $filepath, 'm4v') ? 'video' : 'audio';
```
same thing

AmirRezaM75 commented on App/Vimeo/VimeoDownloader.php:
```diff
@@ -50,11 +50,9 @@ public function download($vimeoId, $filepath)
     private function downloadSource($baseURL, $sourceData, $filepath)
     {
-        file_put_contents($filepath, base64_decode($sourceData['init_segment'], true));
+        file_put_contents($filepath, base64_decode((string) $sourceData['init_segment'], true));
```
same thing

geneowak replied:
this one is an array attribute, is there a way of type hinting that particular array attribute?

AmirRezaM75 replied:
Try this:

```php
/**
     * @param array{init_segment: string} $sourceData
 */
````

AmirRezaM75 commented:
@geneowak It would be good to replace annotations like `@param` with actual types in php.
I will do this if you are not interested

geneowak commented:
> @geneowak It would be good to replace annotations like `@param` with actual types in php. I will do this if you are not interested

could you give me an example?

geneowak committed:
Add type declarations

geneowak committed:
Fix remaining type declarations.

geneowak committed:
add a github pr template.

geneowak commented:
@AmirRezaM75 I've also added a PR template because I think it will make it easier for people making PRs. Feel free to edit it to the project's standards.

AmirRezaM75 commented:
> > @geneowak It would be good to replace annotations like `@param` with actual types in php. I will do this if you are not interested
> 
> could you give me an example?

You already done this.

for example we have authenticate method. which has bunch of `@param`

````
/**
     * Tries to login.
     *
     * @param  string  $email
     * @param  string  $password
     * @return bool
     *
     * @throws LoginException
     */
    public function authenticate($email, $password)
````

we can refactor it to

````php

/**
     * @throws LoginException
     */
public function authenticate(string $email, string $password): bool
````

AmirRezaM75 commented:
> @AmirRezaM75 I've also added a PR template because I think it will make it easier for people making PRs. Feel free to edit it to the project's standards.

ok, please remove `UI/UX improvement`

geneowak committed:
Update pr template and remove remaining @params.

geneowak commented:
> > @AmirRezaM75 I've also added a PR template because I think it will make it easier for people making PRs. Feel free to edit it to the project's standards.
> 
> ok, please remove `UI/UX improvement`

Done, I've also removed the remaining references to params that were not refactored by rector.

geneowak committed:
Fix bug prepareDownloadLink.

geneowak committed:
Fix bug in getSeries by taking advantage of v3 features of the Flysytem.

AmirRezaM75 commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,29 @@
+- [ ] Documentation update
+- [ ] Refactoring
+- [ ] Hotfix
+- [ ] Security patch
```
![Screenshot from 2024-12-15 06-16-52](https://github.com/user-attachments/assets/9406ae10-dc3a-4607-8854-dac53d1d5d47)

It doesn't behave like a radio button. It acts like there are some remaining tasks to be done for the issue

AmirRezaM75 commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,29 @@
+## Title
+[Provide a succinct and descriptive title for the pull request]
```
I think title is redundant.

In general this project is small enough and this template would be overwhelming

AmirRezaM75 commented on App/Downloader.php:
```diff
@@ -17,65 +18,47 @@
-    private $bench;
+    private readonly Ubench $bench;
 
     // [string => number[]]
```
````php
/** @array<string, int[]> */
````

AmirRezaM75 commented on App/Downloader.php:
```diff
@@ -135,41 +118,39 @@ public function start($options)
 
-        if (empty($email) or empty($password))
-            throw new LoginException("No EMAIL and PASSWORD is set in .env file");
+        if ($email === '' || $email === '0' || ($password === '' || $password === '0')) {
```
replace with `empty `

AmirRezaM75 commented on App/Downloader.php:
```diff
@@ -241,40 +222,39 @@ protected function setFilters()
 
             foreach ($episodes as $episode) {
-                $positions = explode(',', $episode);
+                $positions = explode(',', (string) $episode);
```
it's int

geneowak replied:
may be we cast the returned values since explode is a string function.

AmirRezaM75 commented on App/Http/Resolver.php:
```diff
@@ -12,75 +13,45 @@
+         * Guzzle client
+         */
+        private readonly Client $client,
+        /**
+         * Ubench lib
+         */
```
Remove comment

AmirRezaM75 commented on App/Http/Resolver.php:
```diff
@@ -12,75 +13,45 @@
-        $this->bench = $bench;
-        $this->retryDownload = $retryDownload;
+    public function __construct(
+        /**
+         * Guzzle client
+         */
```
Remove comment

AmirRezaM75 commented on App/Http/Resolver.php:
```diff
@@ -231,11 +175,10 @@ private function getLaracastsLink($serieSlug, $episodeNumber)
      * @return string
      */
-    private function getRedirectUrl($url)
+    private function getRedirectUrl(string|array $url): array
```
$url can not be array

AmirRezaM75 commented on App/Http/Resolver.php:
```diff
@@ -290,18 +224,14 @@ private function downloadVideo($downloadUrl, $saveTo)
         $url = $this->getRedirectUrl($url);
-        $url = $this->getRedirectUrl($url);
-        $parts = parse_url($url);
+        $parts = parse_url($url[0]);
```
needs attention

geneowak replied:
get redirectUrl now returns an array because getHeader in guzzle 7 returns and array (https://docs.guzzlephp.org/en/stable/psr7.html#accessing-headers)

geneowak replied:
I think the better way would have been to do `return $response->getHeader('Location')[0];` in the get `redirectUrl` method

AmirRezaM75 replied:
Yeah, I did the same thing

AmirRezaM75 commented on App/System/Controller.php:
```diff
@@ -1,67 +1,63 @@
-        $this->system = $system;
-    }
+    public function __construct(
+        /**
+         * Flysystem lib
+         */
```
Remove comment

AmirRezaM75 commented on App/System/Controller.php:
```diff
@@ -1,67 +1,63 @@
     {
-        $list  = $this->system->listContents(SERIES_FOLDER, true);
-        $array = [];
+        // we want only files, and we only need their paths
+        $list = $this->system->listContents(SERIES_FOLDER, true)
+            ->filter(fn (StorageAttributes $attributes) => $attributes->isFile())
+            ->sortByPath()
+            ->map(fn (StorageAttributes $attrs) => $attrs->path())
+            ->toArray();
```
needs attention

AmirRezaM75 commented on App/System/Controller.php:
```diff
@@ -132,55 +128,51 @@ private function getSkippedData($pathToSkipFile)
 
         return $this->system->has($file)
-            ? require $this->system->getAdapter()->getPathPrefix() . $file
+            ? ''.$file
```
:warning:

geneowak replied:
getPathPrefix was removed from the flysystem (https://github.com/thephpleague/flysystem/issues/1436) but I wasn't sure what this was trying to do so just left the default that the flysystem uses for the path prefix.

AmirRezaM75 replied:
Actually we no longer can access to adapter because of visibility. I used `->read()` method but that was not enough, because we actually trying to import (require) a php file. So I changed that to a json file instead. It's more convenient and takes less storage

AmirRezaM75 commented on App/Utils/SeriesCollection.php:
```diff
@@ -1,46 +1,31 @@
-     * @return $this
-     */
-    public function where($key, $value)
+    public function where(string $key, string $value): \App\Utils\SeriesCollection
```
Replace `\App\Utils\SeriesCollection` with self

AmirRezaM75 reviewed (commented):
@geneowak 
I left some comments to don't forget. Ignore them.

AmirRezaM75 commented:
@geneowak Thanks for contribution
Your MR merged into master
https://github.com/carlosflorencio/laracasts-downloader/pull/158

geneowak commented:
> @geneowak Thanks for contribution Your MR merged into master #158

Happy to help, thanks for the great work you do. I've added some replies to your comments to try to explain some of the refactors.
